---
title: "Exploring SHM"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = F, message = F)
library(readr)
library(dplyr)
library(tidyr)
library(ggplot2)
library(cowplot)
theme_set(theme_cowplot())

source('gene_frequency_functions.R')

# Annotated sequences
annotated_seqs <- read_csv('../processed_data/annotated_seqs.csv')
#annotated_seqs <- read_csv('~/Desktop/v_gene_selection/processed_data/annotated_seqs.csv')

clone_info <- read_csv('../processed_data/clone_info.csv')
#clone_info <- read_csv('~/Desktop/v_gene_selection/processed_data/clone_info.csv')
clone_info <- get_info_from_mouse_id(clone_info) %>%
  mutate(group_controls_pooled = factor(group_controls_pooled, levels = group_controls_pooled_factor_levels))

load('../results/precomputed_gene_freqs.RData')
#load('~/Desktop/v_gene_selection/results/precomputed_gene_freqs_all_seqs.RData')

# Add general clone information to objects containing within-clone mutation frequencies
mutation_freqs_within_clones <- left_join(mutation_freqs_within_clones, clone_info) 
mutation_freqs_within_clones_by_tissue_and_cell_type <- left_join(
  mutation_freqs_within_clones_by_tissue_and_cell_type,
  clone_info
) %>%
  mutate(compartment_cell_type = factor(compartment_cell_type, levels = c('naive','GC','PC','mem','nonnaive_IgD+B220+')),
         compartment_tissue = factor(compartment_tissue, levels = c('LN','spleen','BM')),
         group_controls_pooled = factor(group_controls_pooled, levels = group_controls_pooled_factor_levels))

clone_freqs <- left_join(clone_freqs, clone_info %>% select(mouse_id, clone_id, v_gene))
clone_freqs_by_tissue <- left_join(clone_freqs_by_tissue, clone_info %>% select(mouse_id, clone_id, v_gene))
clone_freqs_by_tissue_and_cell_type <- left_join(clone_freqs_by_tissue_and_cell_type, clone_info %>% select(mouse_id, clone_id, v_gene))

```


Across all clones in all mice (and considering mutation frequencies within the clone as a whole, as opposed to tissue or cell type-specific parts of the clone), most amino acid mutations stay at low frequencies, especially when we remove clones with fewer than 10 sequences


```{r echo = F, fig.width= 10, fig.height = 6}
min_clone_size = 10

pl1 <- mutation_freqs_within_clones %>%
  ggplot(aes(x = mutation_freq_in_compartment)) +
  geom_histogram(aes(y = stat(count) / sum(count)), 
                     binwidth = 0.05) +
  xlab('Mutation frequency within clone') +
  ylab('Fraction of mutations')

pl2 <- mutation_freqs_within_clones %>%
  filter(total_clone_prod_seqs >= min_clone_size) %>%
  ggplot(aes(x = mutation_freq_in_compartment)) +
  geom_histogram(aes(y = stat(count) / sum(count)), 
                     binwidth = 0.05) +
  xlab('Mutation frequency within clone') +
  ylab('Fraction of mutations')

plot_grid(pl1, pl2, labels = c('All (productive) clones', paste0('Clones with at least ', min_clone_size, ' productive seqs.')))
  
```





Does the number of clones with mutations above a certain threshold increase over time? We choose a threshold of 50% and look at the fraction of clones with 0,1,2,...,n mutations at or above that threshold in each mouse (each line). It looks like mice with two infections have more clones with several mutations at or above 50% frequency, but it's hard to see:

```{r echo = F, warning = F}
mutation_freq_threshold <- 0.5

# For each productive clone, counts the number of mutations within the clone at or above a chosen frequency
# If the clone_freqs and mutation_freqs_within_clones objects are not broken down by compartment (cell type / tissue),
# the frequency of each mutation is computed across the entire clone, including sequences from different cell types and tissues
# If clone_freqs and mutation_freqs_within_clones are broken down by compartment, it will count the number of mutations in each clone 
# at or above the threshold specifically in each compartment

# For each clone, counts number of high-frequency mutations
count_mutations_above_threshold_by_clone <- function(clone_freqs, mutation_freqs_within_clones,
                                                   mutation_freq_threshold){
  
  # Clone freqs is used to ensure clones so that clones with no mutations at all (thus missing from mutation_freqs_within_clones) 
  # are counted (as having 0 mutations at above any chosen frequency threshold except 0)
  
  stopifnot(any(clone_freqs$n_clone_seqs_in_compartment == 0) == F)
  
  
  all_productive_clones <- clone_freqs %>%
    select(mouse_id, day, group_controls_pooled, infection_status, clone_id, v_gene, matches('compartment_tissue'),
           matches('compartment_cell_type'), n_clone_seqs_in_compartment)
  
  grouping_vars <- c('mouse_id', 'group_controls_pooled', 'day', 'infection_status', 'clone_id','v_gene')
  
  compartment_vars <- c('compartment_cell_type','compartment_tissue')
  
  if(all(compartment_vars %in% names(clone_freqs))){
    grouping_vars <- c(grouping_vars, compartment_vars)
  }
  
  mutations_above_threshold_by_clone <- mutation_freqs_within_clones %>%
    group_by(across(all_of(grouping_vars))) %>%
    filter(mutation_freq_in_compartment >= mutation_freq_threshold) %>%
    dplyr::summarise(n_mutations_above_threshold = n(),
                     mutations_list = paste(mutation, collapse = ';'))

  mutations_above_threshold_by_clone <- left_join(all_productive_clones, mutations_above_threshold_by_clone) %>%
    mutate(n_mutations_above_threshold = ifelse(is.na(n_mutations_above_threshold), 0,
                                                n_mutations_above_threshold)) %>%
    mutate(mutation_freq_threshold = mutation_freq_threshold) %>%
    select(mutation_freq_threshold, everything())
  
  return(mutations_above_threshold_by_clone)
  
}

mutations_above_threshold_by_clone <- count_mutations_above_threshold_by_clone(clone_freqs,
                                                                             mutation_freqs_within_clones,
                                                                             mutation_freq_threshold)

get_distribution_n_mutations_above_threshold <- function(mutations_above_threshold_by_clone,
                                                                  min_clone_size){
  
  grouping_vars <- c('mutation_freq_threshold', 'mouse_id', 'day', 'group_controls_pooled', 'infection_status',
                     'n_mutations_above_threshold')
  
  compartment_vars <- c('compartment_cell_type', 'compartment_tissue')
  if(all(compartment_vars %in% names(mutations_above_threshold_by_clone))){
    grouping_vars <- c(grouping_vars, compartment_vars)
  }

  mutations_above_threshold_by_clone %>%
    filter(n_clone_seqs_in_compartment >= min_clone_size) %>%
    group_by(across(grouping_vars)) %>%
    dplyr::summarise(n_clones = n()) %>%
    group_by(across(grouping_vars[grouping_vars != 'n_mutations_above_threshold'])) %>%
    mutate(fraction_clones = n_clones / sum(n_clones)) %>%
    ungroup()
  
}

get_distribution_n_mutations_above_threshold(mutations_above_threshold_by_clone,
                                             min_clone_size = min_clone_size) %>%
  filter(n_mutations_above_threshold <= 10) %>%
  ggplot(aes(x = n_mutations_above_threshold, y = fraction_clones,
             color = infection_status)) +
  #geom_col() +
  #geom_point() +
  geom_line(aes(group = mouse_id), alpha = 0.3) +
  geom_smooth(aes(group = infection_status), linetype = 2, se = F) +
  #facet_wrap('group_controls_pooled') +
  scale_y_log10(breaks = c(1,0.5,0.25,0.1,0.05,0.01),
                labels = c(1,0.5,0.25,0.1,0.05,0.01)) +
  xlab('Number of mutations at or above 50% frequency within the clone') +
  ylab(paste0('Fraction of clones with at least\n', min_clone_size, ' productive sequences')) +
  theme(legend.position = c(0.7,0.9)) +
  scale_color_discrete(name = 'Infection status')


```




To more easily visualize these data, we plot, for clones with at least a given number of sequences, what fraction of them have at least a certain number of mutations above the 50% threshold. For instance, for each mouse (each point), we plot the fraction of clones with at least 100 sequences that have at least one mutation at or above 50% threshold. 



```{r echo = F, warning = F}
get_fraction_clones_with_n_mutations_above_threshold <- function(mutations_above_threshold_by_clone,
                                                                 n_mutations_threshold,
                                                                 min_clone_size){
  
  dist_mutations_above_threshold <- get_distribution_n_mutations_above_threshold(
    mutations_above_threshold_by_clone, min_clone_size = min_clone_size
  )
  
  grouping_vars <- c('mutation_freq_threshold', 'mouse_id', 'day', 'group_controls_pooled', 'infection_status')
  
  compartment_vars <- c('compartment_cell_type', 'compartment_tissue')
  if(all(compartment_vars %in% names(dist_mutations_above_threshold))){
    grouping_vars <- c(grouping_vars, compartment_vars)
  }
  
  
  dist_mutations_above_threshold %>%
    group_by(across(grouping_vars)) %>%
    mutate(n_clones_denominator = sum(n_clones)) %>%
    filter(n_mutations_above_threshold >= n_mutations_threshold) %>%
    group_by(across(c(grouping_vars, 'n_clones_denominator'))) %>%
    dplyr::summarise(fraction_clones_with_at_least_n_mutations =
                       sum(fraction_clones),
                     n_clones_with_at_least_n_mutations = sum(n_clones)) %>%
    mutate(n_mutations_threshold = n_mutations_threshold,
           min_clone_size = min_clone_size) %>%
    select(matches('compartment_tissue'), matches('compartment_cell_type'), mutation_freq_threshold, min_clone_size, n_mutations_threshold, everything()) %>%
    ungroup()
  

}

value_combinations <-
  expand.grid(n_mutations_threshold = c(1,2,3),
              min_clone_size = c(0,10,100))


plot_combinations <- mapply(FUN = get_fraction_clones_with_n_mutations_above_threshold,
       n_mutations_threshold = value_combinations$n_mutations_threshold,
       min_clone_size =  value_combinations$min_clone_size,
       MoreArgs = list(mutations_above_threshold_by_clone = mutations_above_threshold_by_clone), 
       SIMPLIFY = F)

plot_combinations <- bind_rows(plot_combinations) %>%
  mutate(n_mutations_threshold = case_when(
    n_mutations_threshold == 1 ~ '1+ mutations >= 50%',
    n_mutations_threshold == 2 ~ '2+ mutations >= 50%',
    n_mutations_threshold == 3 ~ '3+ mutations >= 50%')) %>%
  mutate(min_clone_size = case_when(
    min_clone_size == 0 ~ 'All productive clones',
    min_clone_size == 10 ~ 'Clones with 10+ productive seqs',
    min_clone_size == 100 ~ 'Clones with 100+ productive seqs'
  ))

plot_combinations %>%
  filter(min_clone_size == 'Clones with 100+ productive seqs', n_mutations_threshold == '1+ mutations >= 50%') %>%
  ggplot(aes(x = group_controls_pooled,
             y = fraction_clones_with_at_least_n_mutations,
             color = infection_status)) +
  geom_boxplot(outlier.alpha = 0) +
  geom_point() +
  background_grid() +
  theme(legend.position = 'top', 
        axis.text.x = element_text(size = 8, angle = 40,
                                     vjust = 0.5)) +
  xlab('Group') +
  ylab('Fraction of clones (100+ seqs)\nwith 1+ mutations >= 50%') +
  scale_color_discrete(name = 'Infection status') +
  scale_y_continuous(limits = c(0,NA))

```
We see that, compared with controls, there's a burst of mostly unmutated clones early in the response, with clones accumulating high frequency mutations as the response progresses. This result is consistent with the observation by [Nielsen et al.](https://www.sciencedirect.com/science/article/pii/S1931312820305035?via%3Dihub) that there's an influx of mostly unmutated clones early in the response to SARS-CoV-2, followed by an increase in the average % of sites mutated (they're looking at all mutations, not just mutations above a certain threshold, but still). Like we're doing for the plot above, their analysis is also not constrained to subpopulations likely to be specific for the antigen (they're looking at PBMCs, we're looking at clones from all tissues we sampled). (A similar observation for Ebola virus is made by [Davis et al.](https://www.sciencedirect.com/science/article/pii/S0092867419304556?via%3Dihub)).


We can now do that for several minimum clone sizes and for several number of mutations above the threshold. We find that late in the primary response and in the secondary response, mice have more clones with several mutations at or above 50% frequency, especially among the largest clones.

```{r echo = F, fig.width=8, fig.height=6, warning = F}
plot_combinations %>%
  ggplot(aes(x = group_controls_pooled,
             y = fraction_clones_with_at_least_n_mutations,
             color = infection_status)) +
  geom_boxplot(outlier.alpha = 0) +
  geom_point() +
  facet_grid(n_mutations_threshold~min_clone_size, scales = 'free') +
  background_grid() +
  theme(legend.position = 'top',
        axis.text.x = element_text(size = 8, angle = 40,
                                     vjust = 0.5)) +
  xlab('Group') +
  ylab('Fraction of clones') +
  scale_color_discrete(name = 'Infection status')
  
```


Now we break things down into compartments (tissue / cell type combinations). For each cell type in each tissue, we consider clones with at least 10 sequences in that compartment. 
The number of clones with at least 1 mutation at or above 50% frequency (with the frequency measured relative to number of sequences from the clone in the corresponding compartment) increases over time among GC cells and PCs in the lymph node.


```{r echo = F, fig.width= 6, fig.height=4}

mutations_above_threshold_by_clone_by_compartment <- count_mutations_above_threshold_by_clone(clone_freqs_by_tissue_and_cell_type,
                                                                                              mutation_freqs_within_clones_by_tissue_and_cell_type,
                                                                                              mutation_freq_threshold = 0.5) %>%
   mutate(compartment_cell_type = factor(compartment_cell_type, levels = c('naive','GC','PC','mem','nonnaive_IgD+B220+')), 
          group_controls_pooled = factor(group_controls_pooled, levels = group_controls_pooled_factor_levels),
          compartment_tissue = factor(compartment_tissue, levels = c('LN','spleen','BM'))) %>%
  filter(n_clone_seqs_in_compartment >= 10)


  
fraction_clones_with_1plus_mutations_by_compartment <-
  get_fraction_clones_with_n_mutations_above_threshold(mutations_above_threshold_by_clone_by_compartment,
                                                       n_mutations_threshold = 1, min_clone_size = 10) %>%
    mutate(day = as.integer(as.character(day)))

# Add clone rank in each compartment
mutations_above_threshold_by_clone_by_compartment <- left_join(
  mutations_above_threshold_by_clone_by_compartment,
  clone_freqs_by_tissue_and_cell_type %>% select(mouse_id, clone_id, compartment_tissue, compartment_cell_type,
                                                 clone_rank_in_compartment)
) %>%
   mutate(compartment_cell_type = factor(compartment_cell_type, levels = c('naive','GC','PC','mem','nonnaive_IgD+B220+')), 
          group_controls_pooled = factor(group_controls_pooled, levels = group_controls_pooled_factor_levels),
          compartment_tissue = factor(compartment_tissue, levels = c('LN','spleen','BM')))



# Plots for exporting
fraction_clones_with_high_freq_muts_LN_plot <- fraction_clones_with_1plus_mutations_by_compartment %>%
  filter(compartment_cell_type %in% c('GC','mem','PC'), compartment_tissue == 'LN',
       group_controls_pooled != 'control') %>%
  mutate(compartment_cell_type = case_when(
    compartment_cell_type == 'GC' ~ 'Lymph node GC cells',
    compartment_cell_type == 'PC' ~ 'Lymph node plasma cells',
    compartment_cell_type == 'mem' ~ 'Lymph node memory cells'
  )) %>%
  mutate(compartment_cell_type = factor(compartment_cell_type, levels = c('Lymph node GC cells',
                                                  'Lymph node plasma cells',
                                                  'Lymph node memory cells'))) %>%
  ggplot(aes(x = day,
             y = fraction_clones_with_at_least_n_mutations,
             color = infection_status,
             group = day)) +
  geom_boxplot(outlier.alpha = 0) +
  geom_point(position = position_jitter(width = 0.1), aes(size = n_clones_denominator), alpha = 0.8) +
  background_grid() +
  theme(legend.position = 'top') +
  xlab('Days after primary infection') +
  ylab('Fraction of clones with at least 10 sequences\nthat have mutations at or above 50% frequency') +
  scale_color_manual(values = c('green3','dodgerblue2')) +
  scale_size_continuous(name = 'Number of clones') +
  scale_y_continuous(limits = c(0,NA)) +
  facet_grid(.~compartment_cell_type)  +
  guides(color = 'none') +
  scale_x_continuous(breaks = c(8,16,24,40,56))

fraction_clones_with_high_freq_muts_all_tissues_plot <- fraction_clones_with_1plus_mutations_by_compartment %>%
  filter(compartment_cell_type %in% c('GC','PC','mem')) %>%
  mutate(compartment_cell_type = case_when(
    compartment_cell_type == 'GC' ~ 'Germinal center cells',
    compartment_cell_type == 'PC' ~ 'Plasma cells',
    compartment_cell_type == 'mem' ~ 'Memory cells'
  ),
  compartment_tissue = case_when(
           compartment_tissue == 'LN' ~ 'Mediastinal LN',
           compartment_tissue == 'spleen' ~ 'Spleen',
           compartment_tissue == 'BM' ~ 'Bone marrow'
         )) %>%
  mutate(compartment_cell_type = factor(compartment_cell_type, levels = c('Germinal center cells',
                                                  'Plasma cells',
                                                  'Memory cells')),
         compartment_tissue = factor(compartment_tissue, levels = c('Mediastinal LN',
                                            'Spleen','Bone marrow'))) %>%
  ggplot(aes(x = group_controls_pooled,
             y = fraction_clones_with_at_least_n_mutations,
             color = infection_status)) +
  geom_boxplot(outlier.alpha = 0) +
  geom_point(position = position_jitter(width = 0.1), aes(size = n_clones_denominator), alpha = 0.8) +
  background_grid() +
  theme(legend.position = 'top', 
        axis.text.x = element_text(size = 8, angle = 40,
                                     vjust = 0.5)) +
  xlab('Group') +
  ylab('Fraction of clones with at least 10 sequences\nthat have mutations at or above 50% frequency') +
  scale_color_discrete(name = 'Infection') +
  scale_size_continuous(name = 'Number of clones') +
  scale_y_continuous(limits = c(0,NA)) +
  facet_grid(compartment_tissue~compartment_cell_type)

plot(fraction_clones_with_high_freq_muts_all_tissues_plot)
```
Does the average number of mutations in expanded clones increase over time?

```{r echo =F,  fig.width= 6, fig.height=4}
mean_n_mutations_above_threshold_by_compartment <- mutations_above_threshold_by_clone_by_compartment %>%
  group_by(mouse_id, day, group_controls_pooled, infection_status,
           compartment_tissue, compartment_cell_type) %>%
  dplyr::summarise(mean_n_mutations_above_threshold = mean(n_mutations_above_threshold),
                   n_clones_in_compartment = dplyr::n()) %>%
  ungroup() %>%
  mutate(day = as.integer(as.character(day)))


mean_n_mutations_LN_plot <- mean_n_mutations_above_threshold_by_compartment %>%
  filter(compartment_cell_type %in% c('GC','mem','PC'), compartment_tissue == 'LN',
       group_controls_pooled != 'control') %>%
  mutate(compartment_cell_type = case_when(
    compartment_cell_type == 'GC' ~ 'Lymph node GC cells',
    compartment_cell_type == 'PC' ~ 'Lymph node plasma cells',
    compartment_cell_type == 'mem' ~ 'Lymph node memory cells'
  )) %>%
  mutate(compartment_cell_type = factor(compartment_cell_type, levels = c('Lymph node GC cells',
                                                  'Lymph node plasma cells',
                                                  'Lymph node memory cells'))) %>%
  ggplot(aes(x = day,
             y = mean_n_mutations_above_threshold,
             color = infection_status,
             group = day)) +
  geom_boxplot(outlier.alpha = 0) +
  geom_point(position = position_jitter(width = 0.2), aes(size = n_clones_in_compartment), alpha = 0.9) +
  background_grid() +
  theme(legend.position = 'top') +
  xlab('Days after primary infection') +
  ylab('Average number of mutations at or\nabove 50% frequency in clones with at least 10 seqs.') +
  scale_color_manual(values = c('green3','dodgerblue2')) +
  scale_y_continuous(limits = c(-0.05,NA)) +
  facet_grid(.~compartment_cell_type) +
  scale_size_continuous(name = 'Number of clones') +
  guides(color = 'none') +
  scale_x_continuous(breaks = c(8,16,24,40,56))
  
  

mean_n_mutations_all_tissues_plot <- mean_n_mutations_above_threshold_by_compartment %>%
  filter(compartment_cell_type %in% c('GC','PC','mem')) %>%
  mutate(compartment_cell_type = case_when(
    compartment_cell_type == 'GC' ~ 'Germinal center cells',
    compartment_cell_type == 'PC' ~ 'Plasma cells',
    compartment_cell_type == 'mem' ~ 'Memory cells'
  ),
  compartment_tissue = case_when(
           compartment_tissue == 'LN' ~ 'Mediastinal LN',
           compartment_tissue == 'spleen' ~ 'Spleen',
           compartment_tissue == 'BM' ~ 'Bone marrow'
         )) %>%
  mutate(compartment_cell_type = factor(compartment_cell_type, levels = c('Germinal center cells',
                                                  'Plasma cells',
                                                  'Memory cells')),
         compartment_tissue = factor(compartment_tissue, levels = c('Mediastinal LN',
                                            'Spleen','Bone marrow'))) %>%
  ggplot(aes(x = group_controls_pooled,
             y = mean_n_mutations_above_threshold,
             color = infection_status)) +
  geom_boxplot(outlier.alpha = 0) +
  geom_point(position = position_jitter(width = 0.2), aes(size = n_clones_in_compartment), alpha = 0.9) +
  background_grid() +
  theme(legend.position = 'top', 
        axis.text.x = element_text(size = 8, angle = 40,
                                     vjust = 0.5)) +
  xlab('Group') +
  ylab('Average number of mutations at or\nabove 50% frequency in clones with at least 10 seqs.') +
  scale_color_discrete(name = 'Infection status') +
  scale_y_continuous(limits = c(-0.05,NA)) +
  facet_grid(compartment_tissue~compartment_cell_type) +
  scale_size_continuous(name = 'Number of clones')

plot(mean_n_mutations_all_tissues_plot)

```



Do the most abundant clones in lymph node PC and GC cell populations tend to have more mutations at or above 50% frequency? No, the number of high frequency mutations within a clone does not seem to be associated with the clone's rank

```{r echo = F}

clone_rank_vs_high_freq_muts_LN_PCs_plot <- mutations_above_threshold_by_clone_by_compartment %>%
  filter(compartment_cell_type == 'PC', compartment_tissue == 'LN',
         group_controls_pooled != 'control') %>%
  ggplot(aes(x = clone_rank_in_compartment, y = n_mutations_above_threshold)) +
  geom_point(aes(group = mouse_id, color = infection_status), alpha = 0.2) +
  facet_wrap('group_controls_pooled', nrow = 2, scales = 'free') +
  theme(legend.position = 'none') +
  geom_smooth(se = T, color = 'black', method = 'loess') +
  #scale_y_log10() +
  ylim(-0.05,20) +
  scale_color_manual(values = c('green3','dodgerblue2')) +
  xlab('Clone rank in lymph node plasma cells') +
  ylab('Number of amino acid mutations\nat or above 50% frequency') +
  ggtitle('(Y axis truncated at 20 mutations to improve visualization)')

clone_rank_vs_high_freq_muts_LN_GCs_plot <- mutations_above_threshold_by_clone_by_compartment %>%
  filter(compartment_cell_type == 'GC', compartment_tissue == 'LN',
         group_controls_pooled != 'control') %>%
  ggplot(aes(x = clone_rank_in_compartment, y = n_mutations_above_threshold)) +
  geom_point(aes(group = mouse_id, color = infection_status), alpha = 0.2) +
  facet_wrap('group_controls_pooled', nrow = 2, scales = 'free') +
  theme(legend.position = 'none') +
  geom_smooth(se = T, color = 'black', method = 'loess') +
  #scale_y_log10() +
  ylim(-0.05,20) +
  scale_color_manual(values = c('green3','dodgerblue2')) +
  xlab('Clone rank in lymph node germinal center cells') +
  ylab('Number of amino acid mutations\nat or above 50% frequency') +
  ggtitle('(Y axis truncated at 20 mutations to improve visualization)')
 
```

```{r echo = F}
# EXPORTING PLOT
save(fraction_clones_with_high_freq_muts_LN_plot,
     fraction_clones_with_high_freq_muts_all_tissues_plot,
     mean_n_mutations_LN_plot,
     mean_n_mutations_all_tissues_plot,
     clone_rank_vs_high_freq_muts_LN_PCs_plot,
     clone_rank_vs_high_freq_muts_LN_GCs_plot,
     file = '../figures/all_seqs_freqs/exported_ggplot_objects/high_frequency_mutations.RData')

```




## Parallel evolution in clones using the same V gene ##



```{r echo = F}
# For each pair of mice, find all clones sharing the same V gene. For all pairs of clones, compute the number of high-frequency mutations they have in common
get_n_clones_by_high_freq_mutation <- function(mutation)

# Grouping must be done before calling this function so that only clones sharing the same v gene are compared
# Otherwise shared mutations will be misleading (same mutation means different things in different genes)
count_mutations_shared_by_clone_pairs <- function(mutations_list){
  # Compares every element of mutations_list against the others
  
  if(length(mutations_list) > 1){
      paired_lists <- t(combn(mutations_list, 2))
  colnames(paired_lists) <- c('mutations_list_clone_i', 'mutations_list_clone_j')
  paired_lists <- as_tibble(paired_lists)
  
  # Internal vectorized function.
  compare_mutation_lists <- function(mutations_list_i, mutations_list_j){
    split_list_i <- str_split(mutations_list_i,';')
    split_list_j <- str_split(mutations_list_j,';')
    
    output <- mapply(list_i = split_list_i, list_j = split_list_j,
                               FUN = function(list_i, list_j){
                                 shared_mutations = list_i[list_i %in% list_j]
                                 
                                 return(length(shared_mutations[!is.na(shared_mutations)]))
                                 }
                     )
    return(output)
  }
  
  shared_mutations = paired_lists %>% mutate(shared_mutations = compare_mutation_lists(mutations_list_i = mutations_list_clone_i,
                                                                    mutations_list_j = mutations_list_clone_j)) 
  }else{
    shared_mutations <- tibble(mutations_list_clone_i = c(), mutations_list_clone_j = c(), shared_mutations = c())
  }  
  return(shared_mutations)
  
}
  
shared_mutations_in_LN_clones <-  mutations_above_threshold_by_clone_by_compartment %>% 
  filter(group_controls_pooled != 'control', 
         compartment_tissue == 'LN', compartment_cell_type %in% c('GC','PC','mem')) %>%
  group_by(group_controls_pooled, infection_status, compartment_cell_type, v_gene) %>%
  summarise(count_mutations_shared_by_clone_pairs(mutations_list = mutations_list)) %>%
  ungroup()

shared_mutations_in_LN_clones %>%
  mutate(any_mutations_shared = shared_mutations > 0) %>%
  group_by(group_controls_pooled, compartment_cell_type, any_mutations_shared) %>%
  count() %>%
  group_by(group_controls_pooled, compartment_cell_type) %>%
  mutate(probability = n/sum(n)) %>%
  filter(any_mutations_shared == T) %>%
  ggplot(aes(x = group_controls_pooled, y = probability)) +
  geom_col() +
  facet_wrap('compartment_cell_type', nrow = 1) +
  xlab('group') +
  ylab('Probability that two clones sharing the same V allele\nhave high-frequency mutations in common') +
  ylim(0,1)

shared_mutations_in_LN_clones %>%
  filter(grepl('IGHV14-4', v_gene)) %>%
  mutate(any_mutations_shared = shared_mutations > 0) %>%
  group_by(group_controls_pooled, compartment_cell_type, any_mutations_shared) %>%
  count() %>%
  group_by(group_controls_pooled, compartment_cell_type) %>%
  mutate(probability = n/sum(n)) %>%
  filter(any_mutations_shared == T) %>%
  ggplot(aes(x = group_controls_pooled, y = probability)) +
  geom_col() +
  facet_wrap('compartment_cell_type', nrow = 1) +
  xlab('group') +
  ylab('Probability that two clones sharing the same V allele\nhave high-frequency mutations in common') +
  ylim(0,1)
  



```


